From 7b9e9301e9ea5f82a2d6978b49117b1d37c25601 Mon Sep 17 00:00:00 2001
From: Mateusz Koza <mateusz.koza@grinn-global.com>
Date: Fri, 28 Nov 2025 10:18:51 +0100
Subject: [PATCH] linux mac add support for TXC 90-degree shift

---
 .../ethernet/stmicro/stmmac/dwmac-generic.c   | 26 +++++++++++++++++++
 1 file changed, 26 insertions(+)

diff --git a/drivers/net/ethernet/stmicro/stmmac/dwmac-generic.c b/drivers/net/ethernet/stmicro/stmmac/dwmac-generic.c
index 90f3745ed..13781fd22 100644
--- a/drivers/net/ethernet/stmicro/stmmac/dwmac-generic.c
+++ b/drivers/net/ethernet/stmicro/stmmac/dwmac-generic.c
@@ -16,6 +16,26 @@
 #include "stmmac.h"
 #include "stmmac_platform.h"
 
+#define RGMII_TXC_90DEG_SEL		0xF7EA0A04
+#define RGMII_TXC_90DEG_SEL_SIZE	0x4
+#define RGMII_TXC_90DEG_ENABLE		0x1
+
+static int syna_enable_txc_90deg(struct device *dev)
+{
+	void __iomem *txc_90deg_reg;
+
+	txc_90deg_reg = devm_ioremap(dev, 
+			      RGMII_TXC_90DEG_SEL, 
+			      RGMII_TXC_90DEG_SEL_SIZE);
+	if (!txc_90deg_reg) {
+		dev_err(dev, "Failed to remap RGMII TXC phase shift reg\n");
+		return -ENOMEM;
+	}
+
+	writel(RGMII_TXC_90DEG_ENABLE, txc_90deg_reg);
+	return 0;
+}
+
 static int dwmac_generic_probe(struct platform_device *pdev)
 {
 	struct plat_stmmacenet_data *plat_dat;
@@ -26,6 +46,12 @@ static int dwmac_generic_probe(struct platform_device *pdev)
 	if (ret)
 		return ret;
 
+	if (of_property_read_bool(pdev->dev.of_node, "syna,txc-90deg")) {
+		ret = syna_enable_txc_90deg(&pdev->dev);
+		if (ret)
+			return ret;
+	}
+
 	if (pdev->dev.of_node) {
 		plat_dat = stmmac_probe_config_dt(pdev, stmmac_res.mac);
 		if (IS_ERR(plat_dat)) {
-- 
2.48.1

